### ISSUE TITLE
What CI checks for our repos would you prefer?

### ISSUE THREAD

**Issue by ismay** [dhis2/notes#21]
Just wondering what services everyone prefers for their PR checks/repo badges. Personally my ideal setup would be:

- Travis
- Greenkeeper
- Codecoverage (codecov or coveralls)
- Vulnerability alerts (snyk.io is nice, but github also checks this)

Just wondering what everyone else's ideal setup would be. Hope that this is not out of scope for this repo.

**Comment by amcgee** [comment]
I would like to add some kind of **auto-provisioned deployment preview**.  I've set this up with [netlify](https://netlify.com) for Data-Vis, Maps, and Dashboards (though it's a hassle until we change the way the Core does CORS filtering).  What this allows is:

Visual inspection and smoke testing of an application PR without cloning / starting / pointing at a random server.  It is **very** helpful for QA folks and i.e. @joeDHIS who don't really care how something is implemented, but it's also very useful for devs to get a quick picture of an application after a change and to do regression testing in a consistent environment.

While not _technically_ a CI "check", I think this is important and provides high-level confidence in the safety of a change.

**Comment by ismay** [comment]
I wonder if automated screenshot diffs for components would be nice. I've seen services that provide such things, but have never worked with it.

**Comment by varl** [comment]
The `ui` lib doesn't have a backend dependency so having a visual diff would be kind of handy. Since @joeDHIS is speccing those components would probably be very convenient for him too.

**Comment by amcgee** [comment]
Good call @varl - I've added netlify to the `ui` library, you can see the
master branch at [dhis2-ui.netlify.com](https://dhis2-ui.netlify.com) and
all PRs will also be built and accessible at a custom netlify sub-domain
(linked from the status check in the PR).  There is also a GitHub bot we
could add which would comment on the PR with a link to the deployed version
for testing.

[/end discussion hijack]

**Comment by ismay** [comment]
Nice 🎉🚀, so an updated list would be:

- Travis
- Greenkeeper
- Codecoverage (codecov or coveralls)
- Vulnerability alerts (snyk.io is nice, but github also checks this)
- Netlify to check PR changes in a test env.
- Visual diffs

I guess these checks don't apply in all situations, but as a wishlist for me this seems quite nice.

**Comment by varl** [comment]
I don't have that many checks I think are necessary, though I do have some requirements of the checks:

- **Quick**, so they don't block work.
- **Valuable**, so they are never ignored.
- **Stable**, so they don't produce noise (e.g. false positives/negatives).

Lord Kelvin stated (paraphrase) that _"What gets measured gets managed"_ and I find that there is truth in that statement.

We should only have **PR checks** for the things we want to manage.

As for **badges** ... I don't have a strict opinion on that. They provide at-a-glance info and acts as a cue for certain behaviors in the context of a repository. 

Most importantly, badges need to be relevant and up-to-date.

**Comment by ismay** [comment]
Yeah I agree. Those seem like good guidelines.

**Comment by amcgee** [comment]
One more to add, not applicable everywhere and I'm not sure there's a good existing SaaS for it but monitoring bundle size (and bundle size changes) would be really handy.

I think there could be some utility in presenting visual representations of some non-required stats about a particular PR, like bundle size and netlify link and maybe cypress network traffic volume etc, as a bot-generated comment rather than a full-blown status check?  Runs into some update issues though, don't want a comment for every commit.  Thoughts while thinking.

**Comment by amcgee** [comment]
Agree with @varl that we'll want to limit noise and not have 10 checks running constantly, especially if they're slow and/or not useful (see the license check we had in some repos for a while, which could take 30 minutes and threw false positives all the time)

**Comment by amcgee** [comment]
This may be a tangential discussion (not well-supported in the GitHub Issues format unfortunately):

I stumbled upon [this](https://medium.com/@gajus/distributing-a-self-replicating-malicious-code-using-npm-cf2bf3209293) (2 year old) article which still seems possibly relevant and I think it's worth thinking about for our many apps using greenkeeper and travis for npm publishing.  At the very least we should make sure we're looking at diffs rather than blindly merging greenkeeper PRs.

Also, we may want to think about Travis alternatives if [this](https://news.ycombinator.com/item?id=19218036) trend continues

**Comment by ismay** [comment]
> I stumbled upon this (2 year old) article which still seems possibly relevant and I think it's worth thinking about for our many apps using greenkeeper and travis for npm publishing. At the very least we should make sure we're looking at diffs rather than blindly merging greenkeeper PRs.

Yeah definitely. Actually, that's why I added snyk.io to all my projects a while ago (when the event-stream package was compromised). It'll check dependencies for known vulnerabilities (both pr's and your entire project). I find github vulnerability detection a bit too vague to rely on for this as they don't check PRs and don't notify you via email when something's compromised.

**Comment by ismay** [comment]
> Also, we may want to think about Travis alternatives if this trend continues

I didn't know that, that's not a good development.. As an alternative, I've found https://buddy.works to be quite nice to work with. They integrate well with docker as well, which might be nice for our core docker images. Only downside is that they aren't free.

**Comment by varl** [comment]
> Also, we may want to think about Travis alternatives if [this](https://news.ycombinator.com/item?id=19218036) trend continues

There's always our self-hosted Jenkins...

**Comment by ismay** [comment]
> There's always our self-hosted Jenkins...

😭

**Comment by amcgee** [comment]
> > I stumbled upon this (2 year old) article which still seems possibly relevant and I think it's worth thinking about for our many apps using greenkeeper and travis for npm publishing. At the very least we should make sure we're looking at diffs rather than blindly merging greenkeeper PRs.
> 
> Yeah definitely. Actually, that's why I added [snyk.io](https://snyk.io) to all my projects a while ago (when the event-stream package was compromised). It'll check dependencies for known vulnerabilities (both pr's and your entire project). I find github vulnerability detection a bit too vague to rely on for this as they don't check PRs and don't notify you via email when something's compromised.

I think for that particular vulnerability (leaking secret keys to malicious package updates because greenkeeper creates a local branch with the new dep and allows it to be built by Travis) wouldn't actually be solved with snyk.io - it's an edge case, but worth keeping in mind.

**Comment by ismay** [comment]
> I think for that particular vulnerability (leaking secret keys to malicious package updates because greenkeeper creates a local branch with the new dep and allows it to be built by Travis) snyk.io wouldn't actually solve the problem - it's an edge case, but worth keeping in mind.

Yaeh definitely. I meant more that it reminded me to have something in place that will notify me of any security problems.

**Comment by ismay** [comment]
Speaking of CI checks, the license scan seems to not be activated for certain repos, and failed on others. Does anyone care about having this, or could this be removed as well?

**Comment by varl** [comment]
>Cool! Also a nice bot btw: https://packtracker.io. Checks the size of your webpack bundle, whether the >size increased or not, etc. Easy way to visualize your bundle contents/size online.

_Originally posted by @ismay in https://github.com/dhis2/notes/issues/24#issuecomment-471440603_

>I think there could be some utility in presenting visual representations of some non-required stats about a particular PR, like bundle size [...]

This is a full-blown status check, but it does do this for us.

**Comment by varl** [comment]
Ok, so far:

- [ ] - **Travis**: for builds, tests, etc.
- [ ] - **Netlify**: for feature branch testing
- [ ] - **Semantic Pull Request*: for verifying that the commits and PR title is according to spec

**Comment by ismay** [comment]
Nice. So a summary of the suggestions in this thread:

- [x] Travis
- [x] Greenkeeper (not really a status check)
- [x] Netlify
- [x] Semantic PR
- [ ] packtracker.io
- [ ] Codecov
- [ ] Snyk.io

Anything missing? Do we want the last three ones? I'd vote in favor of the last two. And I wouldn't mind packtracker. Other than that I don't think there's anything missing for me in the list.

**Comment by amcgee** [comment]
@ismay I don't know much about packtracker.io specifically but I think we *need* something like it

**Comment by ismay** [comment]
Currently we're using a size limit on some repos that @varl has set up I believe

**Comment by amcgee** [comment]
> @ismay I don't know much about packtracker.io specifically but I think we _need_ something like it

(also in favor of coverage and vulnerability monitoring)

**Comment by varl** [comment]
Any alternatives to packtracker.io?

This is what I found when setting it up:

- It requires **Webpack**
- It does not support **Rollup**
- It does not support analyzing NPM packages
- It requires manual set up for each project on their website
- Cannot grant privileges on packtracker.io to e.g. a GitHub Org, need to manually administer users

**Comment by varl** [comment]
`size-limit` only gives us a hard stop when something creeps to a certain size.

I'd be OK with a tool that generated a `size-history` file which looks like:

```
# <sha> <bytes>
f1964e3 10000
9dae00b 20000
```

Which then a small tool could look at and compare the size trend. Is anyone aware of something like that?

For PRs we could just report the delta between the commit on the base branch and the last commit on the feature branch as a status check and I think that would be enough to be useful.

We use `source-map-explorer` or `webpack-bundle-analyzer` (built into `size-limit`) to analyze the builds anyway so I don't think we need the GUI of packtracker.io anyway.

**Comment by varl** [comment]
^ This approach would also work for coverage reports IMO with the benefit of the state being tracked **in the repo** and not given away to be gated by some rando's business.

**Comment by varl** [comment]
QA has signaled that they want to standardize on SonarQube for coverage and code smells so introducing a different service for it can introduce discrepancies in the reports.

https://ci.dhis2.org/sonarqube/dashboard?id=maintenance-app

**Comment by ismay** [comment]
> QA has signaled that they want to standardize on SonarQube for coverage and code smells so introducing a different service for it can introduce discrepancies in the reports.

Yeah you're right, I remember Phil mentioning that. I do have to say that from my perspective as a front-end dev working on an app, sonarqube does not seem like it'd integrate that well with my preferred workflow. It's running a bunch of checks that overlap with other tools (checks eslint should handle, vulnerability checks and coverage). The interface is not as easy to scan as the tools I'm used to using for these kinds of checks. And integrating it with the ci checks needs more manual work, beyond just enabling a bot.

Personally I prefer using Github status checks for anything that should get a devs attention while working on a PR. If it's not in the status checks it's easy to miss (unless it fails the build when something's wrong, then I'll go explore of course).

That's my personal preference, I'm curious what everyone else prefers. And we should probably then talk to Phil, see what requirements there are from QA. See if we can meet everyone's needs with our setup.

**Comment by ismay** [comment]
> Any alternatives to packtracker.io?

Tbh, I'm fine with sticking to our current size-limit setup. Packtracker was more something that I thought would be nice if it integrated well. The services that I care most about are coverage reporting and vulnerability monitoring anyway.

Eventually instead of just a size limit we could maybe just have automated Lighthouse testing (or webpagetest.org), with our netlify deploys. That'd be more useful than just bundle analysis anyway.

**Comment by varl** [comment]
@ismay @amcgee Since snyk.io and GitHub alerts for security vulnerabilities are based on the same database (https://cve.mitre.org/) wouldn't the GitHub integration be sufficient? We already have it on all repos. We just don't "care" about it enough to fix the problems.

E.g. a report for a vulnerability with a reference to the CVE ticket:

![2019-03-20-131728_1096x978_scrot](https://user-images.githubusercontent.com/185449/54683735-b7ffcd00-4b12-11e9-8bc6-27a2802a8803.png)

**Comment by ismay** [comment]
> @ismay @amcgee Since snyk.io and GitHub alerts for security vulnerabilities are based on the same database (https://cve.mitre.org/) wouldn't the GitHub integration be sufficient?

Ah, yeah I think it would be. The main thing I like about snyk.io is that it notifies you. As I understand it github doesn't actively send you an email when there's a new vulnerability that affects your code.

**Comment by amcgee** [comment]
I thought I'd gotten a warning email from GitHub before, but can't find it now so maybe that was a dream.  Github vulnerability notifications is fine by me!

**Comment by ismay** [comment]
> It's a setting, it's not on by default.

Oh really? I didn't know. Well in that case using github would be nice! No need for an additional service then

**Comment by varl** [comment]
https://help.github.com/en/articles/choosing-the-delivery-method-for-your-notifications#choosing-the-delivery-method-for-security-alerts-for-vulnerable-dependencies

**Comment by amcgee** [comment]
I hadn't noticed this before but GitHub has also started including vulnerabilities in the git metadata when pushing:

```
git push
Counting objects: 11, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (11/11), done.
Writing objects: 100% (11/11), 1.22 KiB | 1.22 MiB/s, done.
Total 11 (delta 8), reused 0 (delta 0)
remote: Resolving deltas: 100% (8/8), completed with 8 local objects.
remote: 
remote: GitHub found 1 vulnerability on dhis2/dashboards-app's default branch (1 low). To find out more, visit:
remote:      https://github.com/dhis2/dashboards-app/network/alerts
remote: 
To https://github.com/dhis2/dashboards-app
   468b17f..9533887  chore/upgrade-cra -> chore/upgrade-cra
```

**Comment by varl** [comment]
We seem to have reached a reasonable baseline CI checks.

Replace Greenkeeper with Dependabot, and this thread is still pretty accurate.

We are missing a few for code coverage (but then again; we are also missing _a lot_ of tests). If we want help tracking code coverage, leveraging the QA team (SonarQube) makes sense, so we'll try to roll that out after summer.

Closing this issue for now, feel free to re-open if something new develops.

**Comment by ismay** [comment]
One thing I'd love is either codecov or coveralls as a check on our PRs. I'd like to have that information visible directly in the status checks of the PR.

**Comment by ismay** [comment]
I see that in the comment above you mentioned that you don't like introducing a separate service, which I understand. To me anything that's not in the repo status-checks is quite easy to forget though. Could we add the sonarqube coverage data to the github status checks? I don't really care about the service itself, just that the data is there in the PR.

**Comment by varl** [comment]
The QA team has moved from a custom SonarQube instance to SonarCloud, which is a service that can be run with a GitHub Action and shows up as a status check.

We'd still need to set up a config for what we want it to look for (so it doesn't overlap with ESLint, Prettier, etc.).

I've set up a workflow file for it, and here is an example of integrating it into a repo: https://github.com/dhis2/cli-style/pull/276
